File: data_structures/javascript/HashTable.js
//Why use Hash Tables

//Hash tables are an associated array data structure, mapping keys to values
//Allow for (amortized) constant time lookup, insertion and deletion, although the worst case remains O(n)

//When we use a LList at each bucket to prevent overriding values in the case of collisions, it's called chaining
//We can chain with something other than a linked list (e.g. binary search tree) if we think that they chances of collision are very high

//When not to use Hash Tables

//When you don't need to look up things quickly (use a linked list)
//When the keys are integers and are within a small range (use an array)
//Balanced Binary Search Trees can expand more efficiently
//Binary Search Trees allow for faster in-order traversal of keys
//Hash Tables' hashing functions are hard to generate: they can be either unwieldy or inefficient
//If you can tolerate some overlap, a Bloom Filter is more memory efficient for representing a set

class HashNode {
    constructor(key, value, next) {
        this.key = key;
        this.value = value;
        this.next = next || null;
        this.size = 0;
    }
}

class HashTable {
    constructor() {
        this.size = 0;
        this.buckets = [];
        this.numberOfBuckets = this.buckets.length;
    }



    //O(key.length)
    hash(key) {
        let total = 0;
        for (let i = 0; i < key.length; i++) {
            total += key.charCode(i);
        }
        let bucket = total % this.numberOfBuckets;
        return bucket;
    }

    //O(key.length)
    checkKey(key) {
        if (typeof key === "object" || key == null) {
            throw new TypeError('Unallowed key type of object, null or undefined');
        }

        return typeof key === "string" ? key : key.toString();
    }

    //O(key.length + this.size)
    insert(key, value) {

        const keyString = this.checkKey(key);
        const targetBucket = this.hash(keyString);
        let currentNode = this.buckets[targetBucket];

        if (!currentNode) {
            currentNode = new HashNode(keyString, value);
            this.buckets[targetBucket] = currentNode;
            this.size++;
        } else if (currentNode.key === keyString) {
            currentNode.value = value;
        } else {
            while (currentNode.next) {
                if (currentNode.key === keyString) {
                    currentNode.value = value;
                    return;
                }
                currentNode = currentNode.next;
            }
            currentNode.next = new HashNode(keyString, value);
            this.size++;
        }
    }

    //O(this.size)
    get(key) {
        const keyString = this.checkKey(key);
        const targetBucket = this.hash(keyString);
        let currentNode = this.buckets[targetBucket];

        if (!currentNode) {
            return null;
        } else {
            while (currentNode) {
                if (currentNode.key === keyString) {
                    return currentNode.value;
                }
                currentNode = currentNode.next;
            }
            return null;
        }
    }

    //O(number of stored items in all buckets)
    remove(key) {
        const keyString = this.checkKey(key);
        const targetBucket = this.hash(keyString);
        let currentNode = this.buckets[targetBucket];
        let previousNode;

        if (!currentNode) {
            return null;
        } else if (currentNode.key === keyString) {
            nodeToDelete = currentNode;
            this.buckets[targetBucket] = currentNode.next;
            deletedNode = nodeToDelete;
            nodeToDelete = null;
            return deletedNode;
        } else {
            while (currentNode) {
                previousNode = currentNode;
                currentNode = currentNode.next;
                if (currentNode && currentNode.key === keyString) {
                    nodeToDelete = currentNode;
                    previousNode.next = currentNode.next;
                    deletedNode = nodeToDelete;
                    nodeToDelete = null;
                    return deletedNode;
                }
            }
            return null;
        }
    }

    //O(this.size)
    keys() {
        const keysArr = [];
        for (let i = 0; i < this.numberOfBuckets; i++) {
            let currentNode = this.buckets[i];
            while (currentNode) {
                keysArr.push(currentNode.key);
                currentNode = currentNode.next;
            }
        }
        return keysArr;
    }

    //O(this.size)
    values() {
        const valsArr = [];
        for (let i = 0; i < this.numberOfBuckets; i++) {
            let currentNode = this.buckets[i];
            while (currentNode) {
                valsArr.push(currentNode.value);
                currentNode = currentNode.next;
            }
        }
        return valsArr;
    }
}

export default HashTable;

File: data_structures/ruby/hash_table.rb
class HashNode
    def initialize(key, value, next = nil)
        @key = key
        @value = value
        @next = next.nil? ? nil : next
        @size = 0
    end
end

class HashTable
    def initialize(number_of_buckets)
        @size = 0
        @buckets = Array.new(number_of_buckets, nil)
        @number_of_buckets = number_of_buckets
    end

    def hash(key)
        total = 0
        key.to_s.each_char do |char|
            total += char.ord
        end
        bucket = total % @number_of_buckets
        return bucket
    end

    def insert(key, value)
        current_node = @buckets[hash(key)]

        if current_node.nil?
            current_node = HashNode.new(key, value)
            @buckets[target_bucket] = current_node
            @size += 1;
        elsif (current_node.key == key)
            current_node.value = value
        else
            while current_node.next
                if (current_node.key == key)
                    current_node.value = value
                    return;
                end
                current_node = current_node.next
            end
            current_node.next = HashNode.new(key, value)
            @size += 1;
        end
    end

    def get_value_at(key)
        current_node = @buckets[hash(key)]

        if current_node.nil?
            return nil
        else
            until current_node.next.nil?
                if (current_node.key == key)
                    return current_node.value
                end
                current_node = current_node.next
            end
            if current_node.key == key
                return current_node.value
            else
                return nil
            end
        end
    end

    def remove(key)
        target_bucket = hash(key)
        current_node = @buckets[target_bucket]
        previous_node = nil

        if current_node.nil?
            return nil
        elsif (current_node.key == key)
            node_to_delete = current_node
            this.buckets[target_bucket] = current_node.next
            deleted_node = node_to_delete
            node_to_delete = nil
            return deleted_node
        else
            while current_node
                previous_node = current_node
                current_node = current_node.next
                if current_node && current_node.key = key
                    node_to_delete = current_node
                    previous_node.next = current_node.next
                    deleted_node = node_to_delete
                    node_to_delete = nil
                    return deleted_node
                end
            end
            return nil
        end
    end

    def keys
        keys = []
        @buckets.each do |bucket|
            current_node = bucket
            until current_node.nil?
                keys.push(current_node.key)
                current_node = current_node.next
            end
        end
        return keys
    end

    def values
        values = []
        @buckets.each do |bucket|
            current_node = bucket
            until current_node.nil?
                values.push(current_node.value)
                current_node = current_node.next
            end
        end
        return values
    end
end